music21/alpha/analysis/fixer.py

```diff
@@ -316,17 +319,111 @@ def intervalTooBig(self, midiRef, omrRef, setint=5):
             return True
         return False
 
-
 class OrnamentFixer(OMRMidiFixer):
     '''
     Fixes missed ornaments in OMR using expanded ornaments in MIDI
     initialized with self.changes -- a list of tuples in this form:
     (MIDIReference, OMRReference, op)
     MIDIReference and OMRReference are actual note/rest/chord object in some stream
     op is a ChangeOp that relates the two references
+
+    recognizer can be a single recognizer or list of recognizers
+    recognizers take in stream of busy notes and optional stream of simple note(s)
+        and return False or an instance of the ornament recognized
     '''
-    pass
+    def __init__(self, changes, midiStream, omrStream, recognizer):
+        super().__init__(changes, midiStream, omrStream)
+        self.recognizer = recognizer
+
+    def findOrnament(self, busyNotes, simpleNotes) -> Optional[expressions.Ornament]:
+        '''
+        Finds an ornament in busyNotes based from simpleNote
+        using provided recognizers.
+
+        :param busyNotes: stream of notes
+        :param simpleNotes: stream of notes
+
+        When recognizer is a list of recognizers, tries the recognizers in order,
+        returning the result of the first one to successfully recognizes an ornament.
+        '''
+        if isinstance(self.recognizer, list):
+            for r in self.recognizer:
+                ornament = r.recognize(busyNotes, simpleNotes=simpleNotes)
+                if ornament:
+                    return ornament
+        else:
+            ornament = self.recognizer.recognize(busyNotes, simpleNotes=simpleNotes)
+            if ornament:
+                return ornament
+        return None
+
+    def addOrnament(self, selectedNote, ornament, *, show=False) -> bool:
+        '''
+        Adds the ornament to the selectedNote when selectedNote has no ornaments already.
+        :param selectedNote: Note.note to add ornament to
+        :param ornament: Expressions.ornament to add to note
+        :param show: True when note should be colored blue
+        :return: True if added successfully,
+                 False if there was already an ornament on the note and it wasn't added
+        '''
+        if not any(isinstance(e, expressions.Ornament) for e in selectedNote.expressions):
+            selectedNote.expressions.append(ornament)
+            if show:
+                selectedNote.style.color = 'blue'
+            return True
+        return False
+
+    def fix(self: _T, *, show=False, inPlace=True) -> Optional[_T]:
+        '''
+        Corrects missed ornaments in omr stream according to mid stream
+        :param show: Whether to show results
+        :param inPlace: Whether to make changes to own omr stream or
+            return a new OrnamentFixer with changes
+        '''
+        changes = self.changes
+        sa = None
+        omrNotesLabeledOrnament = []
+        midiNotesAlreadyFixedForOrnament = []
+
+        if not inPlace:
```

Comment:
Let's think through this in a bit more depth -- is a OMRFixer of any sort the best place to do Stream Alignment?  I can understand some of the pros and con.   A big pro -- the fixer will know exactly what it needs to hash in order to do its job most efficiently.  A big con -- what is the system that takes two scores and aligns and improves them called?  Is it an XXAligner, a XXFixer, a XXDetector?  Generally one class of system should be running the whole show and I can't figure out at this point which one that is.  I thought it was an Aligner, but now I'm not sure.

Reply:
(that was a comment on the whole system in general, not on the refactor to move out of TrillFixer, which I 100% agree with).

Reply:
Yes agreed to give this some more thought!